---
title: "20241023-miRNA-bowtie vs mirtop"
author: "yufu"
date: "`r Sys.Date()`"
output: rmarkdown::github_document
---
## Aim
This document aims to compare the difference/similarity of two type of miRNA expression estimation method (bowtie and mirtop). \
bowtie: Its basically the 'quick & dirty' way to count, then the pipeline turns this into CPM for visualisation. In contrast, mirtop collapses counts across all iso-miRs for a gene, and reports counts per miR.\
These were previously done in https://cgt.coh.org/haemdata/news/index.html#haemdata-0009010 (nf-core/smrnaseq v2.1.0)

The comparison will be made at not only gene expression correlation but also the state-space. \

## Load Required Libraries
```{r package, cache=TRUE}
# load the pinboard before render the file
# setwd(r'(C:\Users\yufu\OneDrive - City of Hope National Medical Center\Documents\miRNA_AML)')
# library(haemdata)
# library(glmGamPoi)
# use_pinboard("onedrive") #when using local R
# # use_pinboard("devel") #when using R on the apollo server


# loading package
library(haemdata)
library(glmGamPoi)
library(tidyr)
library(edgeR)
library(devtools)
library(ggbiplot)
library(dplyr)
library(compositions)
library(tidyHeatmap)
library(factoextra)
library(openxlsx)
```


## Data Loading and Processing
Load data from haemdata \
bowtie_cpm have been deposited in the haemdata, while mirtop is still raw count and need to be converted here.
```{r load_data, cache=TRUE}
# loading data from haemdata
get_pin_list()

metadata<-get_pin("metadata_mmu.csv", version = "20230907T012227Z-d3324")


cpm.bowtie<-get_pin("mmu_mirna_all_mice_miRBase22_bowtie_cpm.csv", version = "20230907T012024Z-a1f46")
colnames(cpm.bowtie)[1]<-"miRNA"
cpm.bowtie <- cpm.bowtie %>%
  dplyr::select(-miRNA) %>%
  magrittr::set_rownames(cpm.bowtie$miRNA)

counts.mirtop<-get_pin("mmu_mirna_all_mice_miRBase22_mirtop_counts.csv", version = "20230907T012005Z-6a6f3")
counts.mirtop <- counts.mirtop %>%
  dplyr::select(-miRNA) %>%
  magrittr::set_rownames(counts.mirtop$miRNA)


# replace NA to 0
sum(is.na(cpm.bowtie))
cpm.bowtie[is.na(cpm.bowtie)]<-0

sum(is.na(counts.mirtop))
counts.mirtop[is.na(counts.mirtop)]<-0
```


# Convert mirtop raw counts to cpm
Codes from https://github.com/drejom/smrnaseq/blob/master/bin/edgeR_miRBase.r (also cloned from nf-core/smrnaseq v2.1.0)
```{r convert_count_to_cpm, cache=TRUE}
# make mirtop into cpm ================================================
raw_to_cpm<-function(data){
  # Remove genes with 0 reads in all samples
  row_sub = apply(data, 1, function(row) all(row ==0 ))
  # Only subset if at least one sample is remaining
  nr_keep <- sum(row_sub)
  if (nr_keep > 0){
    data<-data[!row_sub,]
  }
  # Also check for colSums > 0, otherwise DGEList will fail if samples have entirely colSum == 0 #Fixes #134
  drop_colsum_zero <- (colSums(data, na.rm=T) != 0) # T if colSum is not 0, F otherwise
  data <- data[, drop_colsum_zero] # all the non-zero columns

  # Normalization
  dataDGE<-DGEList(counts=data,genes=rownames(data))
  o <- order(rowSums(dataDGE$counts), decreasing=TRUE)
  dataDGE <- dataDGE[o,]

  # TMM
  dataNorm <- calcNormFactors(dataDGE)

  # Print normalized read counts to file
  dataNorm_df<-as.data.frame(cpm(dataNorm))
  return(dataNorm_df)

  }

cpm.mirtop<-raw_to_cpm(counts.mirtop)

```


# Extract samples from 2018 cohort
```{r extract_2018_129_sample, cache=TRUE}

# arrange the time point and mouse ID in metadata

metadata$timepoint <- factor(metadata$timepoint, levels = c("T0","T1","T2","T3","T4","T5","T6","T6p5","T7","T8","T9","T10","W0",
                                                                     "W2","W3","W4","W5","W6","W7","W8","W9","W10","W11","W12","W13","W14",
                                                                     "W15","W16","W17","W18","W19","W20","W22","W26","W28","W34","W35","W36",
                                                                     "W40","W41","W48","END"))

metadata$mouse_id <- factor(metadata$mouse_id, levels = c("3334","3335","3336","3338","3339","3340","3341","3342","3346","3349",
                                                        "3350","3357","3368","3370","4309","4311","4321","4324","4329","4419",
                                                        "4433","4436","4443","4501","4506","4510","4534","4535"))

metadata<- metadata %>%
  arrange(mouse_id, timepoint)
# create 2018 cohort list

# miRNA129_ID<-metadata.local$ID[metadata.local$cohort=="AML.miRNA.2018"]
# saveRDS(miRNA129_ID,"miRNA129_ID.rds")
# miRNA129_ID<-readRDS("miRNA129_ID.rds")

# instead of read a local list, make the list directly:
miRNA129_ID<-c(
"COHP_21017","COHP_21023","COHP_21029","COHP_21026","COHP_21030","COHP_21020","COHP_21019","COHP_21032","COHP_21024","COHP_21027",
"COHP_21031","COHP_21021","COHP_21025","COHP_21033","COHP_21028","COHP_21035","COHP_21034","COHP_21036","COHP_21037","COHP_21038",
"COHP_21039","COHP_21040","COHP_21041","COHP_21042","COHP_21044","COHP_21046","COHP_21045","COHP_21047","COHP_21048","COHP_21050",
"COHP_21049","COHP_21051","COHP_21052","COHP_21053","COHP_21054","COHP_21055","COHP_21056","COHP_21057","COHP_21058","COHP_21059",
"COHP_21060","COHP_21061","COHP_21062","COHP_21064","COHP_21065","COHP_21066","COHP_21068","COHP_21067","COHP_21069","COHP_21070",
"COHP_21071","COHP_21072","COHP_21073","COHP_21074","COHP_21075","COHP_21076","COHP_21077","COHP_21078","COHP_21079","COHP_21080",
"COHP_21081","COHP_21082","COHP_21084","COHP_21085","COHP_21087","COHP_21086","COHP_21088","COHP_21089","COHP_21090","COHP_21091",
"COHP_21092","COHP_21093","COHP_21094","COHP_21095","COHP_21096","COHP_21097","COHP_21098","COHP_21100","COHP_21101","COHP_21102",
"COHP_21103","COHP_21104","COHP_21105","COHP_21106","COHP_21107","COHP_21108","COHP_21109","COHP_21111","COHP_21110","COHP_21112",
"COHP_21113","COHP_21114","COHP_21115","COHP_21116","COHP_21117","COHP_21118","COHP_21119","COHP_21120","COHP_21123","COHP_21122",
"COHP_21124","COHP_21125","COHP_21126","COHP_21127","COHP_21128","COHP_21129","COHP_21130","COHP_21131","COHP_21132","COHP_21133",
"COHP_21134","COHP_21135","COHP_21136","COHP_21137","COHP_21138","COHP_21139","COHP_21141","COHP_21144","COHP_21142","COHP_21143",
"COHP_21145","COHP_21146","COHP_21147","COHP_21148","COHP_21149","COHP_21150","COHP_21151","COHP_21153","COHP_21154"
)

metadata.AML.miRNA.2018<-metadata[metadata$library_id %in% miRNA129_ID,]%>%
  drop_na(timepoint)

AML.miRNA.2018.library_id<-metadata.AML.miRNA.2018$library_id

# # intersect miRNA species
# miRNA.intersect<-intersect(row.names(counts.bowtie), row.names(counts.mirtop))
# counts.bowtie<-counts.bowtie[row.names(counts.bowtie) %in% miRNA.intersect,]
# counts.mirtop<-counts.mirtop[row.names(counts.mirtop) %in% miRNA.intersect,]

# extract 2018 from bowtie and re-arrange the order of samples in the matrix to fit the order in metadata
cpm.bowtie.2018<-cpm.bowtie[,which(colnames(cpm.bowtie) %in% AML.miRNA.2018.library_id)]
index<-match(metadata.AML.miRNA.2018$library_id, colnames(cpm.bowtie.2018))
cpm.bowtie.2018<-cpm.bowtie.2018[,index]

# extract 2018 from mirtop and re-arrange the order of samples in the matrix to fit the order in metadata
cpm.mirtop.2018<-cpm.mirtop[,which(colnames(cpm.mirtop) %in% AML.miRNA.2018.library_id)]
index<-match(metadata.AML.miRNA.2018$library_id, colnames(cpm.mirtop.2018))
cpm.mirtop.2018<-cpm.mirtop.2018[,index]

# check colnames are at the same order in bowtie and mirtop
all(colnames(cpm.mirtop.2018) == colnames(cpm.bowtie.2018))
```

# Make correlation heatmaps between bowtie and mirtop
```{r make_correlation, cache=TRUE}

# make the combined_cpm matrix 
cpm.bowtie.2018.2<-cpm.bowtie.2018
cpm.mirtop.2018.2<-cpm.mirtop.2018

cpm.bowtie.2018.2$gene<-row.names(cpm.bowtie.2018.2)
cpm.mirtop.2018.2$gene<-row.names(cpm.mirtop.2018.2)

combined_cpm <- dplyr::left_join(cpm.bowtie.2018.2, cpm.mirtop.2018.2, by="gene", suffix = c(".bowtie", ".mirtop")) |>
  tidyr::drop_na()

mat <- combined_cpm |>
  dplyr::select(-gene) |>
  as.matrix() |>
  magrittr::set_rownames(combined_cpm$gene)


# draw correlation heatmap with dendrogram:
correlate_combined_runs <- function(mat, sample_sheet) {
  mat <- mat
  combined_sample_sheet <- sample_sheet
  
  corr_mat_long <- (cor(compositions::acomp(mat), method = "spearman"))^2 |>
    as_tibble(rownames = "sample1") |>
    pivot_longer(!sample1, names_to = "sample2", values_to = "corr_R^2")

  corr_mat_long<-corr_mat_long%>%
    separate(sample1, into = c(NA,"library_id"), remove = F)
  corr_mat_long$library_id <- paste0("COHP_",corr_mat_long$library_id)


  tidy_combined_data <- dplyr::left_join(
    corr_mat_long,
    combined_sample_sheet,
    by = c("library_id" = "library_id")
  )
  unique(tidy_combined_data$timepoint)
  p <- tidy_combined_data |>
    tidyHeatmap::heatmap(sample1,
                         sample2,
                         `corr_R^2`,
                         scale = "none",
                         palette_value = circlize::colorRamp2(c(-0.01, 0, 1), c("red", "white", "blue"))) |>
    tidyHeatmap::add_tile(mouse_id) |>
    tidyHeatmap::add_tile(timepoint) |>
    tidyHeatmap::add_tile(treatment) |>
    tidyHeatmap::add_tile(sex) |>
    tidyHeatmap::add_tile(sample_weeks)

  # tidyHeatmap::save_pdf(p, paste0("results/", matrix, "_corr_heatmap.pdf"), width = 16L, units = "in")
  
  return(p)
}

# correlation heatmap without dendrogram:
correlate_combined_runs_2 <- function(mat, sample_sheet) {
  mat <- mat
  combined_sample_sheet <- sample_sheet
  
  corr_mat_long <- (cor(compositions::acomp(mat), method = "spearman"))^2 |>
    as_tibble(rownames = "sample1") |>
    pivot_longer(!sample1, names_to = "sample2", values_to = "corr_R^2")
  
  corr_mat_long<-corr_mat_long%>%
    separate(sample1, into = c(NA,"library_id"), remove = F)
  corr_mat_long$library_id <- paste0("COHP_",corr_mat_long$library_id)
  
  
  tidy_combined_data <- dplyr::left_join(
    corr_mat_long,
    combined_sample_sheet,
    by = c("library_id" = "library_id")
  )
  
  p <- tidy_combined_data |>
    tidyHeatmap::heatmap(sample1,
                         sample2,
                         `corr_R^2`,
                         scale = "none",
                         cluster_rows = F,
                         cluster_columns = FALSE,
                         palette_value = circlize::colorRamp2(c(-0.01, 0, 1), c("red", "white", "blue"))) |>
    tidyHeatmap::add_tile(mouse_id) |>
    tidyHeatmap::add_tile(timepoint) |>
    tidyHeatmap::add_tile(treatment) |>
    tidyHeatmap::add_tile(sex) |>
    tidyHeatmap::add_tile(sample_weeks)
  
  # tidyHeatmap::save_pdf(p, paste0("results/", matrix, "_corr_heatmap.pdf"), width = 16L, units = "in")
  
  return(p)
 } 

correlate_combined_runs(mat,metadata.AML.miRNA.2018)
correlate_combined_runs_2(mat,metadata.AML.miRNA.2018)

```

## PCA
cpm were subjected to offset, log transform, then perform PCA by {prcomp}  
```{r PCA, cache=TRUE}

var.offset.t.log<-function(mat){
  # drop those rows with zero var
  mat<-mat[which(apply(mat, 1, var) != 0), ]
  # add offset to the dataset
  mat<-mat+0.01
  # transpose
  mat<-t(mat)
  # make log
  mat<-log(mat,2)
  return(mat)
}


cpm.bowtie.2018.pcocessed<-var.offset.t.log(cpm.bowtie.2018)
cpm.mirtop.2018.pcocessed<-var.offset.t.log(cpm.mirtop.2018)

# for test
a<-cpm.bowtie.2018.pcocessed
meta<-metadata.AML.miRNA.2018
suffix<-"bowtie"

# PCA
PCA_plot<-function(a,meta,suffix, negative){
# perfrom PCA
pc <- prcomp(a,
             center = TRUE,
             scale. = F)
# elbow plot 
elb<-fviz_eig(pc)+
  ggtitle(suffix)

# adjust the orientation
if (negative==T){
  pc$x=-pc$x
  pc$rotation=-pc$rotation
  
}

# PC1 PC2 plot
PC12<-ggbiplot(pc,
              obs.scale = 1,
              var.scale = 1,
              groups = meta[["treatment"]],
              # colour = meta[["treatment"]],
              ellipse = F,
              circle = F,
              ellipse.prob = 0.68,
              var.axes = FALSE,)+
  geom_path(aes(colour = meta[["treatment"]]), arrow = arrow(), group= meta[["mouse_id"]])+
  ggtitle(suffix)


pc.sample.df<-as.data.frame(pc$x) #sample
pc.sample.df$ID<-row.names(pc.sample.df)

pc.mi.df<-as.data.frame(pc$rotation)  #miRNA

# miRNA-PC1 vs PC2 colored by sample
PC12.sample<-ggbiplot(pc,
               obs.scale = 1,
               var.scale = 1,
               groups = meta[["mouse_id"]],
               # colour = meta[["treatment"]],
               ellipse = F,
               circle = F,
               ellipse.prob = 0.68,
               var.axes = FALSE)+
  geom_path(aes(colour = meta[["mouse_id"]]), arrow = arrow(), group= meta[["mouse_id"]])+
  ggtitle(suffix)


print(elb)
print(PC12)
# print(PC12.sample)


aa<-list("pc" = pc,
     "elb" = elb,
     "PC12" = PC12,
     "pc.sample.df"=pc.sample.df,
     "pc.mi.df"=pc.mi.df)

return(aa)

}

set.seed(126)
PCA.bowtie<-PCA_plot(cpm.bowtie.2018.pcocessed,metadata.AML.miRNA.2018,"bowtie",negative = F)
PCA.mirtop<-PCA_plot(cpm.mirtop.2018.pcocessed,metadata.AML.miRNA.2018,"mirtop",negative = T)

```

## Merge PC values from bowtie and mirtop, make plots to demonstrate the PC1 and PC2
```{r other_PC_plot, cache=TRUE}

# merge
total.meta<-merge(metadata.AML.miRNA.2018, PCA.bowtie[["pc.sample.df"]],c(1:2,ncol(PCA.bowtie)), by.x="library_id", by.y = "ID")
total.meta<-merge(total.meta, PCA.mirtop[["pc.sample.df"]],c(1:2,ncol(PCA.mirtop)), by.x="library_id", by.y = "ID", suffixes = c(".bowtie",".mirtop"))
total.meta.sort <- total.meta %>%                 # Order table with dplyr
  arrange(timepoint)

# write table
cohort<-"miRNA-2018-bowtie_vs_mirtop"
file.name<-paste(Sys.Date(),cohort,".xlsx", sep = "-")
print(file.name)
openxlsx::write.xlsx(total.meta.sort,file =file.name)

# miRNA-PC1 vs PC2
ggplot(total.meta.sort, aes(x=PC1.bowtie, y=PC2.bowtie, group=mouse_id, color=mouse_id))+
  geom_point()+
  geom_path(arrow = arrow())+
  ggtitle("PC1 vs PC2-bowtie")

ggplot(total.meta.sort, aes(x=PC1.mirtop, y=PC2.mirtop, group=mouse_id, color=mouse_id))+
  geom_point()+
  geom_path(arrow = arrow())+
  ggtitle("PC1 vs PC2-mirtop")

# miRNA-PC1 vs time
ggplot(total.meta.sort, aes(x=timepoint, y=PC1.bowtie, group=mouse_id, color=mouse_id))+
  geom_point(aes(shape=treatment), size = 2)+
  scale_shape_manual(values=c(16, 1))+
  geom_path(aes(linetype=treatment))+
  scale_y_reverse()+
  theme(axis.text.x = element_text(size=8,angle=45, hjust = 1, vjust = 1))+
  ggtitle("PC1 vs time-bowtie")

ggplot(total.meta.sort, aes(x=timepoint, y=PC1.mirtop, group=mouse_id, color=mouse_id))+
  geom_point(aes(shape=treatment), size = 2)+
  scale_shape_manual(values=c(16, 1))+
  geom_path(aes(linetype=treatment))+
  scale_y_reverse()+
  theme(axis.text.x = element_text(size=8,angle=45, hjust = 1, vjust = 1))+
  ggtitle("PC1 vs time-mirtop")


# miRNA-PC2 vs time
ggplot(total.meta.sort, aes(x=timepoint, y=PC2.bowtie, group=mouse_id, color=mouse_id))+
  geom_point(aes(shape=treatment), size = 2)+
  scale_shape_manual(values=c(16, 1))+
  geom_path(aes(linetype=treatment))+
  scale_y_reverse()+
  theme(axis.text.x = element_text(size=8,angle=45, hjust = 1, vjust = 1))+
  ggtitle("PC2 vs time-bowtie")

ggplot(total.meta.sort, aes(x=timepoint, y=PC2.mirtop, group=mouse_id, color=mouse_id))+
  geom_point(aes(shape=treatment), size = 2)+
  scale_shape_manual(values=c(16, 1))+
  geom_path(aes(linetype=treatment))+
  scale_y_reverse()+
  theme(axis.text.x = element_text(size=8,angle=45, hjust = 1, vjust = 1))+
  ggtitle("PC2 vs time-mirtop")



```

## Make plots to compare bowtie and mirtop
```{r PC_plot_between_bowtiw_mirtop, cache=TRUE}

ggplot(total.meta.sort, aes(x=PC1.bowtie, y=PC1.mirtop, group=mouse_id, color=mouse_id))+
  geom_point()+
  geom_path(arrow = arrow())+
  ggtitle("PC1.bowtie vs PC1.mirtop")

ggplot(total.meta.sort, aes(x=PC2.bowtie, y=PC2.mirtop, group=mouse_id, color=mouse_id))+
  geom_point()+
  geom_path(arrow = arrow())+
  ggtitle("PC2.bowtie vs PC2.mirtop")

# merge the miRNA eigenvalue from bowtie and mirtop
bowtie.eigenvalue<-PCA.bowtie[["pc.mi.df"]][,c(1:2)]
bowtie.eigenvalue$ID<-row.names(PCA.bowtie[["pc.mi.df"]])
mirtop.eigenvalue<-PCA.mirtop[["pc.mi.df"]][,c(1:2)]
mirtop.eigenvalue$ID<-row.names(PCA.mirtop[["pc.mi.df"]])

merge.eigenvalue<-merge(bowtie.eigenvalue, mirtop.eigenvalue, by="ID", suffixes = c(".bowtie",".mirtop"))
all(is.na(merge.eigenvalue))

ggplot(merge.eigenvalue, aes(x=PC1.bowtie, y=PC1.mirtop))+
  geom_point()+
  ggtitle("eigenvalue-PC1.bowtie vs PC1.mirtop")

ggplot(merge.eigenvalue, aes(x=PC2.bowtie, y=PC2.mirtop))+
  geom_point()+
  ggtitle("eigenvalue-PC2.bowtie vs PC2.mirtop")

```

```{r sessioninfo}
sessionInfo()

# render the document by running this code: rmarkdown::render("20241023-bowtie_vs_mirtop-github.Rmd") 
```

