<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • haemdata</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">haemdata</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9013</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../notebooks/" aria-label="notebooks"><span class="fa fa-notebook"></span> Notebooks</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/haemdata.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/genomes.html">Reference genomes</a></li>
    <li><a class="dropdown-item" href="../articles/howto.html">How do I...</a></li>
    <li><a class="dropdown-item" href="../articles/missing_metadata.html">Missing metadata</a></li>
    <li><a class="dropdown-item" href="../articles/preprocessing.html">Preprocessing</a></li>
    <li><a class="dropdown-item" href="../articles/version_control.html">Version control</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cohmathonc/haemdata/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/cohmathonc/haemdata/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9014" id="haemdata-0009014">haemdata 0.0.0.9014<a class="anchor" aria-label="anchor" href="#haemdata-0009014"></a></h2>
<ul><li>add RNAseq for old and young mice (senescence.mRNA.2022)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9013" id="haemdata-0009013">haemdata 0.0.0.9013<a class="anchor" aria-label="anchor" href="#haemdata-0009013"></a></h2>
<ul><li>add 63 non-AML human samples (BM &amp; PBMC)</li>
<li>harmonise metadata_hsa and include non-AML patient samples</li>
<li>update {targets} pipeline to use {crew} for parallelisation</li>
<li>Add note about Kim et al - pan cancer panel, not transcriptome</li>
<li>Add cell-type plot over time to “How do I…”</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9012" id="haemdata-0009012">haemdata 0.0.0.9012<a class="anchor" aria-label="anchor" href="#haemdata-0009012"></a></h2>
<ul><li>add CML blast crisis scRNAseq experiment; 6 mice, 53 samples (“CML.blastcrisis.2023” cohort)</li>
<li>add Notebook rendering to targets pipeline (via build_package() function)</li>
<li>
<a href="https://www.bioconductor.org/packages/release/bioc/html/isomiRs.html" class="external-link">isomiRs</a> object added to targets pipeline. Can publish pin if any interest</li>
<li>fix FLT3 cohort metadata, include 2 rerun samples, remove irrelevant sample</li>
<li>fix incorrect dod (6 CML mice)</li>
<li>add functions to facilitate survival analyses: add_survival_columns(), plot_cohort_survival()</li>
<li>use “code-link: true” in qmd documents</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9011" id="haemdata-0009011">haemdata 0.0.0.9011<a class="anchor" aria-label="anchor" href="#haemdata-0009011"></a></h2>
<ul><li>add notebooks</li>
<li>renamed mirna &amp; aml2022 10X pins</li>
<li>add CML miR-142 KO scRNAseq (LMPP &amp; T cells) 18 samples; RNAseq 25 samples (“mir142ko” cohort)</li>
<li>update AML Cellplex analysis - Cellranger v7.0.0 (includes intron counts)</li>
<li>change sample_id prefix to MHO from PSON</li>
<li>add miRNA sample metadata from <a href="https://github.com/yufu1120" class="external-link">@yufu1120</a></li>
<li>simplify Isilon paths (<code>/net/nfs-irwrsrchnas01/labs</code> &amp; <code>/net/isi-dcnl/ifs/user_data</code> -&gt; <code>/labs</code>)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9010" id="haemdata-0009010">haemdata 0.0.0.9010<a class="anchor" aria-label="anchor" href="#haemdata-0009010"></a></h2>
<ul><li>move target pipeline to <code>./preprocessing</code>
</li>
<li>move _target.R script to _target.qmd</li>
<li>add treatment mice metadata from <a href="https://github.com/luechi" class="external-link">@luechi</a></li>
<li>make 2016 miRNA timepoints 0-offset to match mRNA (miRNA labels are 1-based)</li>
<li>harmonise miRNA for all AML samples (nf-core/smrnaseq v2.1.0)</li>
<li>miRNA fastqs gzip compressed in place, keeping originals</li>
<li>fix mislabelled human FLT3 AML samples, add new samples</li>
<li>retired GRCm38 mRNAseq pipeline; GENCODEm28 is the default mmu reference genome for mRNAseq</li>
<li>add a MHO sample identifier for all mmu samples</li>
<li>add a warning about reproducibility when pins are loaded without setting <code>version</code>
</li>
<li>rename <code>sample</code> to <code>library_id</code> in metadata_mmu</li>
<li>rename <code>project</code> to <code>cohort</code> in metadata_mmu</li>
<li>add <code>ref_dim1</code> &amp; <code>ref_dim2</code> to metadata_mmu, for PC, UMAP or other coordinates from dimensionality reduction</li>
<li>nf-core sample sheets use <code>library_id</code> column for <code>sample</code> column</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9009" id="haemdata-0009009">haemdata 0.0.0.9009<a class="anchor" aria-label="anchor" href="#haemdata-0009009"></a></h2>
<ul><li>fix metadata_mmu</li>
<li>fix pin versions on Onedrive</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9008" id="haemdata-0009008">haemdata 0.0.0.9008<a class="anchor" aria-label="anchor" href="#haemdata-0009008"></a></h2>
<ul><li>freeze on updating GRCm38 pins</li>
<li>include AML scRNAseq 10X hda5 file paths in metadata_mmu</li>
<li>implement update_metadata_mmu() and retire make_metadata_mmu()</li>
<li>fix erroneous AML mouse sample metadata identified by <a href="https://github.com/yufu1120" class="external-link">@yufu1120</a> (DOD)</li>
<li>fix erroneous AML scRNAseq bulk sample labels (PBMC_CKIT)</li>
<li>fix mislabelled scRNAseq 10X samples (previous fix 321c166; now fixed in Cellranger sample sheet)</li>
<li>other small typos and spelling corrections</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9007" id="haemdata-0009007">haemdata 0.0.0.9007<a class="anchor" aria-label="anchor" href="#haemdata-0009007"></a></h2>
<ul><li>add SingleR <code>cell_type</code> and <code>cell_type_fine</code> to Seurat objects/h5ad files</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9006" id="haemdata-0009006">haemdata 0.0.0.9006<a class="anchor" aria-label="anchor" href="#haemdata-0009006"></a></h2>
<ul><li>add ROADMAP.md</li>
<li>add <code>percent_ckit</code>, <code>qc_pass_mapping</code> to metadata_mmu.csv</li>
<li>flag rather than drop mouse samples that fail mapping (<code>qc_pass_mapping</code>)</li>
<li>fix erroneous AML mouse sample metadata identified by Yu-Husan Fu (<a href="https://github.com/yufu1120" class="external-link">@yufu1120</a>)</li>
<li>other small fixes</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9005" id="haemdata-0009005">haemdata 0.0.0.9005<a class="anchor" aria-label="anchor" href="#haemdata-0009005"></a></h2>
<ul><li>add metadata_mmu template for new samples</li>
<li>add pins for “all mice” <code>SummarisedExperiment</code> and csv expression matrix</li>
<li>update AML sample and mouse metadata provided by <a href="https://github.com/yahueikuo" class="external-link">@yahueikuo</a>: Email 2022-9-2 and PSON Teams channel (<code>General|Copy of matadata_mmu_pivoted_AMLmice.YK.xlsx</code>)</li>
<li>add Howto.Rmd</li>
<li>add <code>assay</code> to <code>metadata_mmu</code>
</li>
<li>
<code>sample_sheets</code> now 4-column nf-core format</li>
<li>numerous small fixes</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9004" id="haemdata-0009004">haemdata 0.0.0.9004<a class="anchor" aria-label="anchor" href="#haemdata-0009004"></a></h2>
<ul><li>make <code>auth_type = "device_code"</code> the default for MS365R</li>
<li>added CML mRNA 2022 part 2 (71 samples)</li>
<li>updated vignettes with figs and examples for mRNAseq &amp; scRNAseq</li>
<li>added sample dates from PSON Teams channel (<code>General|AML.Seq.Samples_dates.xlsx</code>), along with computed columns <code>sample_weeks</code>, <code>age_at_end</code>, <code>age_at_start</code>, <code>age_at_sample</code>
</li>
<li>added <code>R-CMD-CHK</code> GitHub Action for CI testing</li>
<li>added RNAseq reads from 90 libraries published by Kim et al 2020 <a href="https://www.nature.com/articles/s41598-020-76933-2" class="external-link">SciRep</a>; <a href="https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJEB27973" class="external-link">PRJEB27973</a>
</li>
<li>streamlined package installation by removing all but the minimal dependencies</li>
<li>moved to {<code>Microsoft365R</code>} and {<code>pins</code>} for storing processed datasets and metadata.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.0.9000" id="haemdata-0009000">haemdata 0.0.0.9000<a class="anchor" aria-label="anchor" href="#haemdata-0009000"></a></h2>
<ul><li>added a <code>NEWS.md</code> file to track changes to the package.</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Denis O’Meally.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

