<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Minimal metadata for mouse samples — metadata_mmu • haemdata</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Minimal metadata for mouse samples — metadata_mmu"><meta name="description" content="Download a template for new samples
ColumnDescription
sample_idUnique identifier with the format MHO_XXXX that identifies a tissue sample with respect to the mouse, tissue type and sample date. Used to match samples across assays.
library_idIntegrative Genomics Core CLN with the format COHP_XXXXX. Matches sequence reads, QC metrics, raw counts and metadata for each sequencing library. Not unique as some libraries are split across multiple sequencing runs.
fastq_1Full path of fastq1, unique in the table
fastq_2Full path of fastq2, unique in the table, or blank for single-end reads
hdf5Full path of h5 file, Cell Rangers's HDF5 Feature-Barcode Matrix Format
strandednessSequencing library protocol (reverse, forward or unstranded)
assayone of mRNA, miRNA, scRNA
mouse_idA unique 3 or 4 digit identifier for each mouse; database held by Kuo lab for AML mice and Zhang lab for CML.
tissueTissue type: PBMC peripheral blood mononuclear cells; BM bone marrow aspirant; BM_CKIT ckit+ flow sorted BM; PBMC_CKIT ckit+ flow sorted PBMC
timepointTimepoint label; formatting is inconsistent across cohorts
treatmentExperimental treatment group
genotypeMouse genotype
sexSex (M/F)
dobDate of birth, YYYY-MM-DD
cohortLogical groups of samples, generally corresponding to experimental cohorts
batchAn arbitrary string to indicate samples in the same sequencing run; unique within cohorts
dodDate of death, YYYY-MM-DD. NA if the mouse survived the experiment.
sample_dateDate the sample was collected, YYYY-MM-DD
percent_ckitPercentage of c-KIT+ cells, measured by flow cytometry (CD117)
sample_weeksTimepoint in weeks (sample_date - min(sample_date)). Post treatment chemo samples begin at week 0, pretreatment samples &amp;lt; 0.
age_at_startAge at start of the 1st sample, in weeks (min(sample_date) - dob)
deadDead or alive at time of sampling; 0 = alive, 1 = dead
ref_dim1reference dimension 1 - typically PC1; can be UMAP or any other dimension reduction coordinate.
ref_dim2reference dimension 2 - typically PC2
qc_pass_mappingmRNA samples only. TRUE if STAR uniquely mapped reads &amp;gt;= mapping_threshold (5% by default), or FALSE if not.


"><meta property="og:description" content="Download a template for new samples
ColumnDescription
sample_idUnique identifier with the format MHO_XXXX that identifies a tissue sample with respect to the mouse, tissue type and sample date. Used to match samples across assays.
library_idIntegrative Genomics Core CLN with the format COHP_XXXXX. Matches sequence reads, QC metrics, raw counts and metadata for each sequencing library. Not unique as some libraries are split across multiple sequencing runs.
fastq_1Full path of fastq1, unique in the table
fastq_2Full path of fastq2, unique in the table, or blank for single-end reads
hdf5Full path of h5 file, Cell Rangers's HDF5 Feature-Barcode Matrix Format
strandednessSequencing library protocol (reverse, forward or unstranded)
assayone of mRNA, miRNA, scRNA
mouse_idA unique 3 or 4 digit identifier for each mouse; database held by Kuo lab for AML mice and Zhang lab for CML.
tissueTissue type: PBMC peripheral blood mononuclear cells; BM bone marrow aspirant; BM_CKIT ckit+ flow sorted BM; PBMC_CKIT ckit+ flow sorted PBMC
timepointTimepoint label; formatting is inconsistent across cohorts
treatmentExperimental treatment group
genotypeMouse genotype
sexSex (M/F)
dobDate of birth, YYYY-MM-DD
cohortLogical groups of samples, generally corresponding to experimental cohorts
batchAn arbitrary string to indicate samples in the same sequencing run; unique within cohorts
dodDate of death, YYYY-MM-DD. NA if the mouse survived the experiment.
sample_dateDate the sample was collected, YYYY-MM-DD
percent_ckitPercentage of c-KIT+ cells, measured by flow cytometry (CD117)
sample_weeksTimepoint in weeks (sample_date - min(sample_date)). Post treatment chemo samples begin at week 0, pretreatment samples &amp;lt; 0.
age_at_startAge at start of the 1st sample, in weeks (min(sample_date) - dob)
deadDead or alive at time of sampling; 0 = alive, 1 = dead
ref_dim1reference dimension 1 - typically PC1; can be UMAP or any other dimension reduction coordinate.
ref_dim2reference dimension 2 - typically PC2
qc_pass_mappingmRNA samples only. TRUE if STAR uniquely mapped reads &amp;gt;= mapping_threshold (5% by default), or FALSE if not.


"><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">haemdata</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9013</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../notebooks/" aria-label="notebooks"><span class="fa fa-notebook"></span> Notebooks</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/haemdata.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/genomes.html">Reference genomes</a></li>
    <li><a class="dropdown-item" href="../articles/howto.html">How do I...</a></li>
    <li><a class="dropdown-item" href="../articles/missing_metadata.html">Missing metadata</a></li>
    <li><a class="dropdown-item" href="../articles/preprocessing.html">Preprocessing</a></li>
    <li><a class="dropdown-item" href="../articles/version_control.html">Version control</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cohmathonc/haemdata/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Minimal metadata for mouse samples</h1>
      <small class="dont-index">Source: <a href="https://github.com/cohmathonc/haemdata/blob/HEAD/R/metadata.R" class="external-link"><code>R/metadata.R</code></a></small>
      <div class="d-none name"><code>metadata_mmu.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="../metadata_mmu_template.xlsx">Download</a> a template for new samples</p><table class="table table"><tr><td>Column</td><td>Description</td></tr><tr><td><code>sample_id</code></td><td>Unique identifier with the format <code>MHO_XXXX</code> that identifies a tissue sample with respect to the mouse, tissue type and sample date. Used to match samples across assays.</td></tr><tr><td><code>library_id</code></td><td>Integrative Genomics Core <em>CLN</em> with the format <code>COHP_XXXXX</code>. Matches sequence reads, QC metrics, raw counts and metadata for each sequencing library. Not unique as some libraries are split across multiple sequencing runs.</td></tr><tr><td><code>fastq_1</code></td><td>Full path of fastq1, unique in the table</td></tr><tr><td><code>fastq_2</code></td><td>Full path of fastq2, unique in the table, or blank for single-end reads</td></tr><tr><td><code>hdf5</code></td><td>Full path of h5 file, Cell Rangers's HDF5 Feature-Barcode Matrix Format</td></tr><tr><td><code>strandedness</code></td><td>Sequencing library protocol (<code>reverse</code>, <code>forward</code> or <code>unstranded</code>)</td></tr><tr><td><code>assay</code></td><td>one of <code>mRNA</code>, <code>miRNA</code>, <code>scRNA</code></td></tr><tr><td><code>mouse_id</code></td><td>A unique 3 or 4 digit identifier for each mouse; database held by <a href="mailto:YKuo@coh.org?subject=Question%20about%20AML%20mice%20from%20PSON">Kuo lab</a> for AML mice and <a href="mailto:YKuo@coh.org?subject=Question%20about%20AML%20mice%20from%20PSON">Zhang lab</a> for CML.</td></tr><tr><td><code>tissue</code></td><td>Tissue type: <code>PBMC</code> peripheral blood mononuclear cells; <code>BM</code> bone marrow aspirant; <code>BM_CKIT</code> ckit+ flow sorted <code>BM</code>; <code>PBMC_CKIT</code> ckit+ flow sorted <code>PBMC</code></td></tr><tr><td><code>timepoint</code></td><td>Timepoint label; formatting is inconsistent across cohorts</td></tr><tr><td><code>treatment</code></td><td>Experimental treatment group</td></tr><tr><td><code>genotype</code></td><td>Mouse genotype</td></tr><tr><td><code>sex</code></td><td>Sex (<code>M</code>/<code>F</code>)</td></tr><tr><td><code>dob</code></td><td>Date of birth, YYYY-MM-DD</td></tr><tr><td><code>cohort</code></td><td>Logical groups of samples, generally corresponding to experimental cohorts</td></tr><tr><td><code>batch</code></td><td>An arbitrary string to indicate samples in the same sequencing run; unique within cohorts</td></tr><tr><td><code>dod</code></td><td>Date of death, YYYY-MM-DD. <code>NA</code> if the mouse survived the experiment.</td></tr><tr><td><code>sample_date</code></td><td>Date the sample was collected, YYYY-MM-DD</td></tr><tr><td><code>percent_ckit</code></td><td>Percentage of c-KIT+ cells, measured by flow cytometry (CD117)</td></tr><tr><td><code>sample_weeks</code></td><td>Timepoint in weeks (<code>sample_date - min(sample_date)</code>). Post treatment chemo samples begin at week 0, pretreatment samples &lt; 0.</td></tr><tr><td><code>age_at_start</code></td><td>Age at start of the 1st sample, in weeks (<code>min(sample_date) - dob</code>)</td></tr><tr><td><code>dead</code></td><td>Dead or alive at time of sampling; <code>0</code> = alive, <code>1</code> = dead</td></tr><tr><td><code>ref_dim1</code></td><td>reference dimension 1 - typically PC1; can be UMAP or any other dimension reduction coordinate.</td></tr><tr><td><code>ref_dim2</code></td><td>reference dimension 2 - typically PC2</td></tr><tr><td><code>qc_pass_mapping</code></td><td><code>mRNA</code> samples only. <code>TRUE</code> if STAR uniquely mapped reads &gt;= <code>mapping_threshold</code> (5% by default), or <code>FALSE</code> if not.</td></tr></table></div>


    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The <a href="https://github.com/cohmathonc/haemdata/blob/main/scripts/import_metadata.R#L24" class="external-link"><code>update_metadata_mmu()</code></a>
function assembles the metadata for all RNAseq libraries from AML and CML mice, by consolidating
data scraped from multiple sequencing run sheets, directly from sequencing folders, emails, and so forth.
The code is complex and ugly and undoubtedly some errors will have made it through.</p>
<p>Raise an <a href="https://github.com/cohmathonc/haemdata/issues" class="external-link">issue on GitHub</a>
to report erroneous or missing records.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Denis O'Meally</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Denis O’Meally.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

