<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run nf-core/rnaseq pipeline on Apollo — run_nf_core_rnaseq • haemdata</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run nf-core/rnaseq pipeline on Apollo — run_nf_core_rnaseq"><meta name="description" content="Makes an sbatch job script for the nf-core/rnaseq pipeline and submits it to the Apollo cluster.
The pipeline version is specified by the rnaseq_release parameter in the _targets.R file. The default parameters are used for the pipeline, in addition to --pseudo_aligner salmon and --deseq2_vst.
Makes an sbatch job script for the nf-core/smrnaseq pipeline and submits it to the Apollo cluster.
The pipeline version is specified by the smrnaseq_release parameter in the _targets.R file. The default parameters are used for the pipeline.
Makes an sbatch job script for cutadapt and submits it to the Apollo cluster, one job per library."><meta property="og:description" content="Makes an sbatch job script for the nf-core/rnaseq pipeline and submits it to the Apollo cluster.
The pipeline version is specified by the rnaseq_release parameter in the _targets.R file. The default parameters are used for the pipeline, in addition to --pseudo_aligner salmon and --deseq2_vst.
Makes an sbatch job script for the nf-core/smrnaseq pipeline and submits it to the Apollo cluster.
The pipeline version is specified by the smrnaseq_release parameter in the _targets.R file. The default parameters are used for the pipeline.
Makes an sbatch job script for cutadapt and submits it to the Apollo cluster, one job per library."><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">haemdata</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9013</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../notebooks/" aria-label="notebooks"><span class="fa fa-notebook"></span> Notebooks</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/haemdata.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/genomes.html">Reference genomes</a></li>
    <li><a class="dropdown-item" href="../articles/howto.html">How do I...</a></li>
    <li><a class="dropdown-item" href="../articles/missing_metadata.html">Missing metadata</a></li>
    <li><a class="dropdown-item" href="../articles/preprocessing.html">Preprocessing</a></li>
    <li><a class="dropdown-item" href="../articles/version_control.html">Version control</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cohmathonc/haemdata/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Run nf-core/rnaseq pipeline on Apollo</h1>
      <small class="dont-index">Source: <a href="https://github.com/cohmathonc/haemdata/blob/HEAD/R/rnaseq.R" class="external-link"><code>R/rnaseq.R</code></a>, <a href="https://github.com/cohmathonc/haemdata/blob/HEAD/R/smrnaseq.R" class="external-link"><code>R/smrnaseq.R</code></a></small>
      <div class="d-none name"><code>run_nf_core_rnaseq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Makes an <code>sbatch</code> job script for the nf-core/rnaseq pipeline and submits it to the Apollo cluster.
The pipeline version is specified by the <code>rnaseq_release</code> parameter in the <code>_targets.R</code> file. The <a href="https://nf-co.re/rnaseq/3.7/usage" class="external-link">default parameters</a> are used for the pipeline, in addition to <code>--pseudo_aligner salmon</code> and <code>--deseq2_vst</code>.</p>
<p>Makes an <code>sbatch</code> job script for the nf-core/smrnaseq pipeline and submits it to the Apollo cluster.
The pipeline version is specified by the <code>smrnaseq_release</code> parameter in the <code>_targets.R</code> file. The <a href="https://nf-co.re/smrnaseq/2.1.0/usage" class="external-link">default parameters</a> are used for the pipeline.</p>
<p>Makes an <code>sbatch</code> job script for cutadapt and submits it to the Apollo cluster, one job per library.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_nf_core_rnaseq</span><span class="op">(</span><span class="va">run_folder</span>, <span class="va">sample_sheet</span>, <span class="va">ref_genome</span>, qc <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">run_nfcore_smrnaseq</span><span class="op">(</span></span>
<span>  <span class="va">run_folder</span>,</span>
<span>  <span class="va">sample_sheet</span>,</span>
<span>  clip_r1_3p <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"mmu"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">run_cutadapt_smrna</span><span class="op">(</span><span class="va">sample_sheet</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-run-folder">run_folder<a class="anchor" aria-label="anchor" href="#arg-run-folder"></a></dt>
<dd><p>subfolder in which to run nextflow, keeps the pipeline
from overwriting previous runs and enables the <code>-resume</code> feature of Nextflow.</p></dd>


<dt id="arg-sample-sheet">sample_sheet<a class="anchor" aria-label="anchor" href="#arg-sample-sheet"></a></dt>
<dd><p>nf-core sample sheet with the columns <code>sample_id, library_id, fastq_1</code></p></dd>


<dt id="arg-ref-genome">ref_genome<a class="anchor" aria-label="anchor" href="#arg-ref-genome"></a></dt>
<dd><p>nf-core reference genome, see <a href="../articles/genomes.html">Reference Genomes</a> for supported references</p></dd>


<dt id="arg-qc">qc<a class="anchor" aria-label="anchor" href="#arg-qc"></a></dt>
<dd><p>run the full QC pipeline with STAR_Salmon, or Salmon pseudoalignment only, skipping all QC steps (default: TRUE)</p></dd>


<dt id="arg-clip-r-p">clip_r1_3p<a class="anchor" aria-label="anchor" href="#arg-clip-r-p"></a></dt>
<dd><p>number of bases to trim from reads.</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>mmu or hsa. If <em>mmu</em>, additionally sets the 3' adapter to TCTGGAATTCTCGGGTGCCAAGGAACTCC. For <em>hsa</em>, the 3' adapter is auto-discovered.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a path to the run script, or the multiqc report if it exists.</p>
<p>a path to the run script, or the multiqc report if it exists.</p>
<p>a data.frame with the columns <code>library_id, fastq_1</code>, ready to be used as input to the nf-core pipeline</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Execution of the pipeline is independent of the package build script, allowing asynchronous development.
If no <code>run_folder</code> exists, it will be created along with a <code>sample_sheet.csv</code> and <code>run_{ref_genome}.sh</code>,
which is submitted via <code>sbatch</code>. A successful run is detected by the presence of a a multiqc report.
An existing run script wont be overwritten by this function so it must be deleted manually.</p>
<p>Execution of the pipeline is independent of the package build script, allowing asynchronous development.
If no <code>run_folder</code> exists, it will be created along with a <code>sample_sheet.csv</code> and <code>run_{ref_genome}.sh</code>,
which is submitted via <code>sbatch</code>. A successful run is detected by the presence of a a multiqc report.
An existing run script wont be overwritten by this function so it must be deleted manually.</p>
<p>Execution of the pipeline is independent of the package build script, allowing asynchronous development.
If no <code>run_folder</code> exists, it will be created along with a <code>sample_sheet.csv</code> and <code>run_{ref_genome}.sh</code>,
which is submitted via <code>sbatch</code>. A successful run is detected by the presence of a a multiqc report.
An existing run script wont be overwritten by this function so it must be deleted manually.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Denis O'Meally</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Denis O’Meally.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

